pallets/subtensor/src/subnets/leasing.rs

```diff
@@ -0,0 +1,382 @@
+                    .saturating_add(T::DbWeight::get().writes(55_u64))
+                    .saturating_add(T::DbWeight::get().writes(2_u64.saturating_mul(k)))
+                    .saturating_add(Weight::from_parts(0, 2579).saturating_mul(k)),
+            )
+            .into())
+        } else {
+            // We have the max number of contributors, so we don't need to refund anything
+            Ok(().into())
+        }
+    }
+
+    /// Terminate a lease.
+    ///
+    /// The beneficiary can terminate the lease after the end block has passed and get the subnet ownership.
+    /// The subnet is transferred to the beneficiary and the lease is removed from storage.
+    pub fn do_terminate_lease(
+        origin: T::RuntimeOrigin,
+        lease_id: LeaseId,
+        hotkey: T::AccountId,
+    ) -> DispatchResultWithPostInfo {
+        let who = ensure_signed(origin)?;
+        let now = frame_system::Pallet::<T>::block_number();
+
+        // Ensure the lease exists and the beneficiary is the caller
+        let lease = SubnetLeases::<T>::get(lease_id).ok_or(Error::<T>::LeaseDoesNotExist)?;
+        ensure!(
+            lease.beneficiary == who,
+            Error::<T>::ExpectedBeneficiaryOrigin
+        );
+
+        // Ensure the lease has an end block and we are past it
+        let end_block = lease.end_block.ok_or(Error::<T>::LeaseHasNoEndBlock)?;
+        ensure!(now >= end_block, Error::<T>::LeaseHasNotEnded);
+
+        // Transfer ownership to the beneficiary
+        ensure!(
+            Self::coldkey_owns_hotkey(&lease.beneficiary, &hotkey),
+            Error::<T>::BeneficiaryDoesNotOwnHotkey
+        );
+        SubnetOwner::<T>::insert(lease.netuid, lease.beneficiary.clone());
+        Self::set_subnet_owner_hotkey(lease.netuid, &hotkey);
+
+        // Stop tracking the lease coldkey and hotkey
+        let _ = frame_system::Pallet::<T>::dec_providers(&lease.coldkey).defensive();
+        let _ = frame_system::Pallet::<T>::dec_providers(&lease.hotkey).defensive();
+
+        // Remove the lease, its contributors and accumulated dividends from storage
+        let clear_result =
+            SubnetLeaseShares::<T>::clear_prefix(lease_id, T::MaxContributors::get(), None);
+        AccumulatedLeaseDividends::<T>::remove(lease_id);
+        SubnetLeases::<T>::remove(lease_id);
+
+        // Remove the beneficiary proxy
+        T::ProxyInterface::remove_lease_beneficiary_proxy(&lease.coldkey, &lease.beneficiary)?;
+
+        Self::deposit_event(Event::SubnetLeaseTerminated {
+            beneficiary: lease.beneficiary,
+            netuid: lease.netuid,
+        });
+
+        if clear_result.unique < T::MaxContributors::get() {
+            // We have cleared less than the max number of shareholders, so we need to refund the difference
+            let k = clear_result.unique.into();
+            Ok(Some(
+                Weight::from_parts(56_635_122, 6148)
```

Comment:
Did you consider having a separate weight function? It seems cleaner to update it.

Reply:
Yes, I thought about having this like we normally do with the weights.rs file but didn't know where to put this tbh, would probably put this in the leaseing.rs file to be used from dispatches and when refunding, thanks!

Reply:
Fixed